### PR TITLE
Tool script to run before running MRI patch from PR #3327

### PR DESCRIPTION
**Pull request by cmadjar**
This pull request adds a tool script that need to be run before running the MRI patch sent in PR #3327.

This script will look for orphan entries of the following pairs of `table` => `foreign key`:
- `MRICandidateErrors` => `TarchiveID`
- `mri_violations_log` => `TarchiveID`
- `files` => `TarchiveSource`
- `files_qcstatus` => `FileID`
- `mri_upload` => `TarchiveID`
- `mri_upload` => `SessionID`
- `mri_protocol_checks` => `ScanType`
- `tarchive` => `SessionID`

Take the example of table `files_qcstatus` that links to the `files` table using the `FileID` foreign key. If there is an entry for `FileID` 3 in `files_qcstatus` but not in the `files` table, [this SQL patch](SQL/Archive/19.0/2017-11-29-Add_foreign_key_constraint_to_MRI_violation_tables.sql) from PR #3327 will fail creating the `FileID` foreign key due to this entry with no corresponding `FileID` in the table of origin. This applies to all pairs of `table` => `foreign key` listed above.

Therefore, before running the SQL patch that will create the missing foreign keys on the MRI side, we need to make that no orphan entries can be found in the tables listed above. This script takes care of this. It will:
- fetch orphan entries in those tables
- print orphan entries in the terminal
- ask whether the user want to delete those entries (after proper backup) or update their foreign key to be NULL (the update becomes interesting when dealing with files_qcstatus table as the correct FileID can be repopulated based on the SeriesUID and EchoTime of the file)
- depending on the user's answer ('y' or 'n')
> - if 'y', backup the orphan entries using mysqldump into files `backup_release_19-0_upgrade.sql` in the current directory
> - if 'n', update the foreign key value to NULL for the orphan entries  

Here is a screenshot of part of the script's output:
screen shot 2018-01-22 at 3 45 47 pm

Here is a screenshot of the output when no orphan were found:
screen shot 2018-01-22 at 3 49 07 pm

### CODE REVIEW

**Comment by cmadjar** [comment]
@driusan I assigned this to you. Any feedback is welcome :)

**Comment by MounaSafiHarab** [comment]
@cmadjar 
I will test this first thing tomorrow! thanks for the script.

**Comment by MounaSafiHarab** [comment]
@cmadjar 
I did a quick round of testing; it did detect the issues I had in my files_qcstatus table well. So this is great, but it did delete the entries although the .sql file dump did not work (see screenshot below).
So it makes sense to check that the .sql backup can be written before actually deleting the entries.
(I think in my case, maybe because the config.xml has a regular user as opposed to an admin credentials)

tocecile

**Comment by cmadjar** [comment]
@MounaSafiHarab Thank you for testing it!!

1. I will check that the mysqldump worked before moving forward and deleting the entries (I actually thought about this in the middle of the night, haha)

2. I am not sure to understand why adding a FK on SessionID returns an error about the DoB... Not sure what to do for this

3. the reason it is interactive is so that users can decide to delete entries in tarchive and keep the entries in files_qcstatus to remap it afterward with the correct FileID. If we have only one 'y' or 'n' at the command line, it would mean we would treat all the tables the same way which may not make sense for some projects. What do you think? @ridz1208 will having an interactive script be really problematic for the sys admin? Let me know.

**Comment by nicolasbrossard** [comment]
Just a comment regarding the fact that this script is interactive (and the problems it might cause to sysadmins). You can use the Unix 'yes' command to automate the answers you would give to a script, thereby making  it non-interactive. Not sure if this helps, but here it is anyways.

**Comment by cmadjar** [comment]
@MounaSafiHarab I fixed 1. Now it will exit the script if the mysqldump command is not successful. 
Let me know if you have the same issue. Thanks!!

**Comment by cmadjar** [comment]
@nicolasbrossard really cool! Thanks!! Did not know that UNIX command

@MounaSafiHarab if the sys admin don't want the interactive options, just need to type:
yes | php cleanup_mri_tables_for_19-0_release.php

**Comment by MounaSafiHarab** [comment]
I re-tested this; and it is working well. Here are my comments:

1)  I had to run the `tools/DB_date_zeros_removal.php` script before running this script in this pull request (I know the 000-00-00 entries have nothing to do with your script, but databases that have those entries in dates fields and are on MySQL 5.7 won't be able to run this:()

2) the interactive question; it works for me using yes |; so @ridz1208 we cna automate this for the sys admins; 

3) I had an issue running the mysql dump command; I finally figured out it is my database credentials. I am not sure if this needs to be addressed, but for my case, I had to escape the password in your script before I could run it; OR remove the password argument altogether from the mysqldump command in your script and wait for the script to prompt me for it, then it is all good. @driusan, @ridz1208 not sure if it is worth it to have the script escape the password field?

otherwise, all good on my end.

**Comment by cmadjar** [comment]
@MounaSafiHarab Cool! Thanks for testing it!!! Here are my answers:

1. Maybe in the release notes we could tell users that are running MySQL 5.7 to run the `tools/DB_date_zeros_removal.php` before running the script?

2. cool :)

3. Maybe we could document as well in the release notes that before running the script, users must make sure that their DB user in the `config.xml` file has the credential to run `mysqldump`? At least, now the script exits on `mysqldump` failures before deleting anything.

**Comment by MounaSafiHarab** [comment]
> Maybe we could document as well in the release notes that before running the script, users must make sure that their DB user in the config.xml file has the credential to run mysqldump? At least, now the script exits on mysqldump failures before deleting anything.

just to clarify this last point: my DB credentials are correct (otherwise I can not access the front end nor run insertion scripts which require creation of temporary tables etc...). The problem is that the password has special characters which your script does not escape, so running the mysqldump command fails. But yes, if the mysqldump command failed, no deletion happens, so this is good.

File: tools/cleanup_mri_tables_for_19-0_release.php [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    // grep database connection information from NDB_Config for mysqldump
+    $config = NDB_Config::singleton();
+    $config->load();
+    $database = $config->getSettingFromXML("database");
+
+    // create the mysqldump query to back the orphan entries to be deleted
+    $sqldump = "mysqldump " .
+        "-u " . $database['username'] . " " .
+        "-h " . $database['host']     . " " .
+        "-p"  . $database['password'] . " " .
```
Comment (driusan) [inline tools/cleanup_mri_tables_for_19-0_release.php]: These need to be shell escaped. See https://secure.php.net/manual/en/function.escapeshellarg.php

**Comment by driusan** [comment]
@MounaSafiHarab Are there any more problems on your end in this PR or is this good to be merged?

**Comment by MounaSafiHarab** [comment]
> @MounaSafiHarab Are there any more problems on your end in this PR or is this good to be merged?

I will tets it one more time now (but it  was all good anyway before except for the escaping of the password)

**Comment by MounaSafiHarab** [comment]
re-tested and works well!

just failing Travis now

**Comment by driusan** [comment]
I just want to point out that this has a great pull request message/description.

**Pull request merged**